File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/TwoSum.java
package com.only4.algorithm.version4j.leetcode;

import java.util.HashMap;
import java.util.Map;

/**
 * @author zhenyu.jiang
 */
public class TwoSum {
    public int[] twoSum(int[] nums, int target) {
        // 哈希表：数值 -> 索引
        Map<Integer, Integer> numIndexMap = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            int currentNum = nums[i];
            int complement = target - currentNum;

            // 检查补数是否已存在
            if (numIndexMap.containsKey(complement)) {
                return new int[]{numIndexMap.get(complement), i};
            }

            // 存储当前数值和索引
            numIndexMap.put(currentNum, i);
        }

        throw new IllegalArgumentException("No two sum solution");
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/twoSum.kt
package com.only4.algorithm.version4k.leetcode

fun twoSum(nums: IntArray, target: Int): IntArray {
    // 哈希表：数值 -> 索引
    val numIndexMap = HashMap<Int, Int>()

    nums.forEachIndexed { index, currentNum ->
        val complement = target - currentNum

        // 检查补数是否已存在，如果存在则返回结果
        numIndexMap[complement]?.let { complementIndex ->
            return intArrayOf(complementIndex, index)
        }

        // 存储当前数值和索引
        numIndexMap[currentNum] = index
    }

    throw IllegalArgumentException("No two sum solution")
}
